src/doc/trpl/concurrency.md

```diff
@@ -0,0 +1,343 @@
+
+### `Send`
+
+The first type we're going to talk about is
+[`Send`](std/marker/trait.Send.html). When a type `T` implements `Send`, it indicates
+to the compiler that something of this type is able to have ownership transferred
+safely between threads.
+
+This is important to enforce certain restrictions. For example, if we have a
+channel connecting two threads, we would want to be able to send some data
+down the channel and to the other thread. Therefore, we'd implement `Send` on
+that type.
+
+In the opposite way, if we were wrapping a library with FFI that isn't
+threadsafe, we wouldn't want to implement `Send`, and so the compiler will help
+us enforce that it can't leave the current thread.
+
+### `Sync`
+
+The second of these two types is called [`Sync`](std/marker/trait.Sync.html).
+When a type `T` implements `Sync`, it indicates to the compiler that something
+of this type has no possibility of introducing memory unsafety when passing
+`&T`s between threads.
+
+For example, sharing immutable data with a reference count is threadsafe.
+Therefore, this type would implement `Sync`, so that it could be safely
+shared between threads.
+
+These two types allow you to use the type system to make strong guarantees
+about the properties of your code under concurrency. Before we demonstrate
+why, we need to learn how to create a concurrent Rust program in the first
+place!
+
+## Threads
+
+Rust's standard library provides a library for 'threads', which allow you to
+run Rust code in parallel. Here's a basic example of using `Thread`:
+
+```
+use std::thread::Thread;
+
+fn main() {
+    Thread::scoped(|| {
+        println!("Hello from a thread!");
+    });
+}
+```
+
+The `Thread::scoped()` method accepts a closure, which is executed in a new
+thread. It's called `scoped` because this thread returns a join guard:
+
+```
+use std::thread::Thread;
+
+fn main() {
+    let guard = Thread::scoped(|| {
+        println!("Hello from a thread!");
+    });
+
+    // guard goes out of scope here
+}
+```
+
+When `guard` goes out of scope, it will block execution until the thread is
+finished. If we didn't want this behaviour, we could use `Thread::spawn()`:
+
+```
+use std::thread::Thread;
+use std::io::timer;
+use std::time::Duration;
+
+fn main() {
+    Thread::spawn(|| {
+        println!("Hello from a thread!");
+    });
+
+    timer::sleep(Duration::milliseconds(50));
+}
+```
+
+We need to `sleep` here because when `main()` ends, it kills all of the
+running threads.
+
+[`scoped`](std/thread/struct.Builder.html#method.scoped) has an interesting
+type signature:
+
+```text
+fn scoped<'a, T, F>(self, f: F) -> JoinGuard<'a, T>
+    where T: Send + 'a,
+          F: FnOnce() -> T,
+          F: Send + 'a
+```
+
+Specifically, `F`, the closure that we pass to execute in the new thread.
+It has two restrictions: It must be a `FnOnce` from `()` to `T`. This allows
+us to take ownership of the new thread's environment. The other restriction is
```

Comment:
"This allows us to take ownership of the new thread's environment" should be: "This allows the new thread to take ownership of any data it mentions from the parent thread."